base/intfuncs.jl

```diff
@@ -195,6 +195,11 @@ end
 ^(x::Number, p::Integer)  = power_by_squaring(x,p)
 ^(x, p::Integer)          = power_by_squaring(x,p)
 
+# x^p for any literal integer p is lowered to x^Val{p},
+# to enable compile-time optimizations specialized to p.
+# However, we still need a fallback that calls the general ^:
+^{p}(x, ::Type{Val{p}}) = x^p
```

Comment:
Unitful will definitely want to write its own `^(x, Val{p})` methods that are type-stable anyway, however, so in this case the ambiguity warning is helpful.